src/bin/mdbook-xgettext.rs

```diff
@@ -222,4 +497,172 @@ mod tests {
+
+        for msg in catalog.messages() {
+            assert!(!msg.is_translated());
+        }
+
+        assert_eq!(
+            catalog
+                .messages()
+                .map(|msg| (msg.source(), msg.msgid()))
+                .collect::<Vec<(&str, &str)>>(),
+            &[
+                ("src/SUMMARY.md:1", "The Foo Chapter"),
+                ("src/SUMMARY.md:2", "The Bar Section"),
+                ("src/SUMMARY.md:3", "The Baz Subsection"),
+                ("src/foo.md:1", "How to Foo"),
+                ("src/foo.md:3", "The first paragraph about Foo."),
+                ("src/foo/bar.md:1", "How to Bar"),
+                ("src/foo/bar.md:3", "The first paragraph about Bar."),
+                ("src/foo/bar/baz.md:1", "How to Baz"),
+                ("src/foo/bar/baz.md:3", "The first paragraph about Baz."),
+            ]
+        );
+
+        Ok(())
+    }
+
+    // Helper constants written as functions.
+    // Strings cannot be created using const.
+    fn summary_file() -> String {
+        String::from("src/SUMMARY.md")
+    }
+    fn summary_content() -> Vec<String> {
+        vec![
+            String::from("The Foo Chapter"),
+            String::from("The Bar Section"),
+            String::from("The Baz Subsection"),
+        ]
+    }
+    fn foo_file() -> String {
+        String::from("src/foo.md")
+    }
+    fn foo_content() -> Vec<String> {
+        vec![
+            String::from("How to Foo"),
+            String::from("The first paragraph about Foo."),
+        ]
+    }
+    fn bar_file() -> String {
+        String::from("src/foo/bar.md")
+    }
+    fn bar_content() -> Vec<String> {
+        vec![
+            String::from("How to Bar"),
+            String::from("The first paragraph about Bar."),
+        ]
+    }
+    fn baz_file() -> String {
+        String::from("src/foo/bar/baz.md")
+    }
+    fn baz_content() -> Vec<String> {
+        vec![
+            String::from("How to Baz"),
+            String::from("The first paragraph about Baz."),
+        ]
+    }
+
+    #[test_case(1, vec![
+        (summary_file(), summary_content()),
+        (foo_file(), [foo_content(), bar_content(), baz_content()].concat()),
+    ]; "depth_1")]
+    #[test_case(2, vec![
+        (summary_file(), summary_content()),
+        (foo_file(), foo_content()),
+        (bar_file(), [bar_content(), baz_content()].concat()),
+    ]; "depth_2")]
+    #[test_case(3, vec![
+        (summary_file(), summary_content()),
+        (foo_file(), foo_content()),
+        (bar_file(), bar_content()),
+        (baz_file(), baz_content())
+    ]; "depth_3")]
+    #[test_case(4, vec![
+        (summary_file(), summary_content()),
+        (foo_file(), foo_content()),
+        (bar_file(), bar_content()),
+        (baz_file(), baz_content())
+    ]; "depth_greater_than_necessary")]
+    fn test_split_catalog_nested_directories(
+        split_depth: i64,
+        expected_message_tuples: Vec<(String, Vec<String>)>,
```

Comment:
If the problem is that `&str` doesn't work here instead of `String`, then perhaps `&'static str` works?

In general, `&str` is a borrowed string slice: someone has allocated string data and a `&str` refers to this. There is an ellided lifetime here, so the type is actually `&'a str` for some lifetime `'a` (often a generic parameter to the function where the type is a parameter).

When you use a string literal, the Rust compiler is the one allocating the string data (somewhere in the text segment of your binary). All string literals have the type `&'static str`, with a special `'static` lifetime, signifying that they are valid for the entire program duration and never go out of scope.

Reply:
I was able to re-work the expected values to use the `&'static str` - I had trouble trying to see if I could do the same with mapping the Catalog messages to msgids. Do you know if it's possible? The compiler tells me that the Catalog does not live long enough (looking at the call to msgids_from_catalog, where I turn that fn into returning a `Vec<&'static str>`

Reply:
My advice would be to not use `test_case`. Just create a very simple unit test with static strings. Avoid abstracting too much in test — use string literals everywhere to make the test very clear.

Basically, apply the opposite of the normal DRY principle :smile:

Reply:
Split out into multiple tests using the same "book"!